### PR TITLE
[native] Re-factor http client to run http unit tests in http and https mode.

### PR DESCRIPTION
**Pull request by amitkdutta**
We re-factor http client to take https releated information, instead of reading those from system configuration. Note that, https configuration is still passed via system config, just Presto Server and Exchange source reads it and pass to client library. We also extend existing http tests to run in both mode http and https.

### CODE REVIEW

File: presto-native-execution/presto_cpp/main/PrestoServer.cpp [hunk 1]
```diff
@@ -160,7 +160,7 @@ void PrestoServer::run() {
   int httpPort{0};
   int httpExecThreads{0};
 
-  std::string certPath, keyPath, ciphers;
+  std::string certPath, keyPath, ciphers, clientCertAndKeyPath;
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/PrestoServer.cpp:163]: One variable per line.
```
std::string certPath;
std::string keyPath;
...
```

File: presto-native-execution/presto_cpp/main/PrestoExchangeSource.cpp [hunk 2]
```diff
@@ -56,11 +56,15 @@ PrestoExchangeSource::PrestoExchangeSource(
     const folly::Uri& baseUri,
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/PrestoExchangeSource.cpp:56]: Can we cover https in AnnouncerTest.cpp as well? We could do this in followup. Thanks!

File: presto-native-execution/presto_cpp/main/PrestoExchangeSource.cpp [hunk 3]
```diff
@@ -319,10 +325,21 @@ PrestoExchangeSource::createExchangeSource(
     int destination,
     std::shared_ptr<exec::ExchangeQueue> queue,
     memory::MemoryPool* pool) {
-  if (strncmp(url.c_str(), "http://", 7) == 0 ||
-      strncmp(url.c_str(), "https://", 8) == 0) {
+  if (strncmp(url.c_str(), "http://", 7) == 0) {
     return std::make_unique<PrestoExchangeSource>(
         folly::Uri(url), destination, queue, pool);
+  } else if (strncmp(url.c_str(), "https://", 8) == 0) {
+    auto systemConfig = SystemConfig::instance();
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/PrestoExchangeSource.cpp:332]: nit: const these local variables.

File: presto-native-execution/presto_cpp/main/PrestoExchangeSource.cpp [hunk 4]
```diff
@@ -319,10 +325,21 @@ PrestoExchangeSource::createExchangeSource(
...
+    auto clientCertAndKeyPath =
+        systemConfig->httpsClientCertAndKeyPath().value_or("");
+    auto ciphers = systemConfig->httpsClientCertAndKeyPath().value_or("");
+    return std::make_unique<PrestoExchangeSource>(
+        folly::Uri(url),
+        destination,
+        queue,
+        pool,
+        clientCertAndKeyPath,
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/PrestoExchangeSource.cpp:341]: Can we cover the https in PrestoExchangeSourceTest.cpp in followup? Thanks!

File: presto-native-execution/presto_cpp/main/http/tests/HttpTest.cpp [hunk 5]
```diff
@@ -35,6 +35,9 @@ int main(int argc, char** argv) {
 }
 
 namespace {
+
+enum Protocol : int8_t { http = 0, https = 1 };
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/http/tests/HttpTest.cpp:39]: We could use a boolean for this?

File: presto-native-execution/presto_cpp/main/http/tests/HttpTest.cpp [hunk 6]
```diff
@@ -182,9 +185,22 @@ class HttpClientFactory {
   std::unique_ptr<http::HttpClient> newClient(
       const folly::SocketAddress& address,
       const std::chrono::milliseconds& timeout,
+      Protocol protocol,
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/http/tests/HttpTest.cpp:188]:
```
bool useHttps
```

File: presto-native-execution/presto_cpp/main/http/tests/HttpTest.cpp [hunk 7]
```diff
@@ -563,6 +602,11 @@ TEST_F(HttpTest, testReportOnBodyStatsFunc) {
   wrapper.stop();
 }
 
+INSTANTIATE_TEST_CASE_P(
+    HTTPTest,
+    HttpTestFixture,
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/http/tests/HttpTest.cpp:607]:
```
s/HttpTestFixture/HttpTestSuite/
```

**Review by xiaoxmeng — COMMENTED** [review]
@amitkdutta nice change % minors. Thanks!

**Review by majetideepak — COMMENTED** [review]
Please update the commit title and description as that is what gets reflected.
Follow https://www.freecodecamp.org/news/how-to-write-better-git-commit-messages/ for description content.
The https config is still read from the system config right?

```
Add HTTPS support to HttpClient. 
Extend HTTP tests to run in both HTTP and HTTPS modes.
```

**Comment by amitkdutta** [comment]
> Please update the commit title and description as that is what gets reflected. Follow https://www.freecodecamp.org/news/how-to-write-better-git-commit-messages/ for description content. The https config is still read from the system config right?
> 
> Possible commit message
> 
> ```
> Add HTTPS support to HttpClient. 
> Extend HTTP tests to run in both HTTP and HTTPS modes.
> ```

Thanks @majetideepak for the review. I will update the commit message, thanks so much for the suggestion. Yeah it reads from system config where we specify key and certificate paths. The tests provide a very good example on how to set it up, as it works with such key and certificate files. I plan write a note on how to set it up with java coordinator (as we have done within Meta).

File: presto-native-execution/presto_cpp/main/http/tests/HttpTest.cpp (hunk 5 above)
Comment (amitkdutta) [reply presto-native-execution/presto_cpp/main/http/tests/HttpTest.cpp:39]: Initially I was thinking to use boolean, later moved with enum just for readability. Let me move back to boolean, I think that's clear as well.

File: presto-native-execution/presto_cpp/main/PrestoExchangeSource.cpp (hunk 2 above)
Comment (amitkdutta) [reply presto-native-execution/presto_cpp/main/PrestoExchangeSource.cpp:56]: Yeah. I will change Announer and Prestoexchange tests to run in both mode. I need to refactor test code a bit to do so. Will send in the next PR.

File: presto-native-execution/presto_cpp/main/PrestoExchangeSource.cpp (hunk 4 above)
Comment (amitkdutta) [reply presto-native-execution/presto_cpp/main/PrestoExchangeSource.cpp:341]: Yeah I will do in the next PR

File: presto-native-execution/presto_cpp/main/http/HttpClient.cpp [hunk 8]
```diff
@@ -221,27 +225,31 @@ class ConnectionHandler : public proxygen::HTTPConnector::Callback {
...
       : responseHandler_(responseHandler),
         sessionPool_(sessionPool),
         timer_(timer),
         eventBase_(eventBase),
-        address_(address) {}
+        address_(address),
+        clientCertAndKeyPath_(clientCertAndKeyPath),
+        ciphers_(ciphers) {}
+
+  bool useHttp() {
```
[the remark this answers is not in the document]
Comment (amitkdutta) [reply presto-native-execution/presto_cpp/main/http/HttpClient.cpp:239]: I found the negation a bit hard to read. That's why went with useHttp() route.

**Comment by amitkdutta** [comment]
"Meta Internal-Only Changes Check" fails because I imported this diff to test with Meta internal infrastructure.

File: presto-native-execution/presto_cpp/main/http/HttpClient.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (xiaoxmeng) [reply presto-native-execution/presto_cpp/main/http/HttpClient.cpp:239]: So we don't require clientCertAndKeyPath_ and ciphers_ both to be set? Thanks!

File: presto-native-execution/presto_cpp/main/http/HttpClient.h [hunk 9]
```diff
@@ -128,6 +130,13 @@ class HttpClient {
   const folly::HHWheelTimer::UniquePtr timer_;
   const std::function<void(int)> reportOnBodyStatsFunc_;
   const uint64_t maxResponseAllocBytes_;
+  // clientCertAndKeyPath_ points to a file (usually with pem extension) which
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/http/HttpClient.h:133]:
```
s/clientCertAndKeyPath_ points/Points/
```

**Review by xiaoxmeng — APPROVED** [review]
@amitkdutta thanks for the update. Looks nice!

File: presto-native-execution/presto_cpp/main/http/HttpClient.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (amitkdutta) [reply presto-native-execution/presto_cpp/main/http/HttpClient.cpp:239]: Yes, we either need both set or both unset. Added in constructor

`VELOX_CHECK_EQ(clientCertAndKeyPath_.empty(), ciphers_.empty());`

**Review by xiaoxmeng — APPROVED** [review]
@amitkdutta thanks for the update!

**Pull request merged**